docs/streams/developer-guide.html

```diff
@@ -267,21 +295,532 @@
+
+               <code>flatMapValues</code> is preferable to <code>flatMap</code> because it will not cause data re-partitioning. However,
+               it does not allow you to modify the key or key type like <code>flatMap</code> does.
+
+               <pre class="brush: java;">
+               // Split a sentence into words.
+               KStream&lt;byte[], String&gt; sentences = ...;
+               KStream&lt;byte[], String&gt; words = sentences.flatMapValues(value -> Arrays.asList(value.split("\\s+")));
+
+               // Java 7 example: cf. `mapValues` for how to create `ValueMapper` instances
+               </pre>
+           </td>
+       </tr>
+       <tr>
+           <td><b>Foreach</b>: <code>KStream->void or KTable->void </code></td>
+           <td>
+               Terminal operation. Performs a stateless action on each record.<p>
+
+               Note on processing guarantees: Any side effects of an action (such as writing to external systems)
+               are not trackable by Kafka, which means they will typically not benefit from Kafka's processing guarantees.
+
+               <pre class="brush: java;">
+               KStream&lt;String, Long&gt; stream = ...;
+
+               // Print the contents of the KStream to the local console.
+               // Java 8+ example, using lambda expressions
+               stream.foreach((key, value) -> System.out.println(key + " => " + value));
+
+               // Java 7 example
+               stream.foreach(
+                   new ForeachAction&lt;String, Long&gt;() {
+                       @Override
+                       public void apply(String key, Long value) {
+                         System.out.println(key + " => " + value);
+                       }
+               });
+               </pre>
+           </td>
+       </tr>
+       <tr>
+           <td><b>GroupByKey</b>: <code>KStream->KGroupedStream </code></td>
+           <td>
+               Groups the records by the existing key.<p>
+
+               Grouping is a prerequisite for aggregating a stream or a table and ensures that data is properly
+               partitioned ("keyed") for subsequent operations.<p>
+
+               <b>When to set explicit serdes</b>: Variants of <code>groupByKey</code> exist to override the configured default serdes of
+               your application, which you must do if the key and/or value types of the resulting <code>KGroupedStream</code> do
+               not match the configured default serdes.<p>
+
+               <b>Note:</b>
+               Grouping vs. Windowing: A related operation is windowing, which lets you control how to "sub-group" the
+               grouped records of the same key into so-called windows for stateful operations such as windowed aggregations
+               or windowed joins.<p>
+
+               Causes data re-partitioning if and only if the stream was marked for re-partitioning. <code>groupByKey</code> is
```

Comment:
"Causes data re-partitioning if and only if the stream was marked for re-partitioning." should be in bold font.  Same for other instances of "Causes ...".